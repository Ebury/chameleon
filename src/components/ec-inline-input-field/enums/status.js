export const EDITING = 'editing';
export const LOADING = 'loading';
export const READ_ONLY = 'read-only';
