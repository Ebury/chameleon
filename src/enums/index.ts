export enum SortDirection {
  ASC = 'asc',
  DESC = 'desc',
}

export enum SortDirectionCycle {
  LOWEST_FIRST = 1,
  HIGHEST_FIRST = 2,
}

export enum KeyCode {
  ESC = 27,
  ARROW_UP = 38,
  ARROW_DOWN = 40,
}

export enum ZIndexLevel {
  NOTIFICATION = 'notification',
  MODAL = 'modal',
  TOOLTIP = 'tooltip',
  LEVEL1 = 'level-1',
  LEVEL2 = 'level-2',
  LEVEL3 = 'level-3',
}
