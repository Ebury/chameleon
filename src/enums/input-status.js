export const EDIT = 'edit';
export const LOADING = 'loading';
export const READ_ONLY = 'read-only';
