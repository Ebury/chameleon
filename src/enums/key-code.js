export const ESC = 27;
