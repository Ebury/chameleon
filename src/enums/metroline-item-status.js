export const NEXT = 'next';
export const ACTIVE = 'active';
export const COMPLETED = 'completed';
