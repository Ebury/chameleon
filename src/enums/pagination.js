export const PAGE_SIZES = [5, 10, 50, 100];
export const DEFAULT_PAGE_SIZE = 10;
