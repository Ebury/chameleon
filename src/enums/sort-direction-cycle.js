import { SortDirection } from './sort-direction';

export const LOWEST_FIRST = [SortDirection.ASC, SortDirection.DESC];
export const HIGHEST_FIRST = [SortDirection.DESC, SortDirection.ASC];
