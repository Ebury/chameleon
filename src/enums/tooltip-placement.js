export const AUTO = 'auto';
export const TOP = 'top';
export const BOTTOM = 'bottom';
export const LEFT = 'left';
export const RIGHT = 'right';
