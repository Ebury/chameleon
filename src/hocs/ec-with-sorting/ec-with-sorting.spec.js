import { mount, createLocalVue, createWrapper } from '@vue/test-utils';
import * as SortDirection from '../../enums/sort-direction';
import withSorting from './ec-with-sorting';

describe('EcWithSorting', () => {
  function mountEcWithSorting(props, mountOpts) {
    const localVue = createLocalVue();

    const Component = localVue.extend({
      props: {
        sorts: {
          type: Array,
          default: () => [],
        },
      },
      render(h) {
        return h('div');
      },
    });

    const hocWrapper = mount(withSorting(Component), {
      localVue,
      propsData: { ...props },
      ...mountOpts,
    });

    const componentWrapper = createWrapper(hocWrapper.vm.$children[0].$vnode);

    return { hocWrapper, componentWrapper };
  }

  it('should render properly', () => {
    const { hocWrapper } = mountEcWithSorting();
    expect(hocWrapper.element).toMatchSnapshot();
  });

  describe('multi sort', () => {
    function mountMultiEcWithSorting(props, mountOpts) {
      return mountEcWithSorting({ multiSort: true, ...props }, mountOpts);
    }

    it('should use default sort direction when column is not sorted yet', () => {
      const { componentWrapper } = mountMultiEcWithSorting();
      componentWrapper.vm.$emit('sort', 'test-column');
      expect(componentWrapper.vm.sorts).toEqual([
        { column: 'test-column', direction: SortDirection.ASC },
      ]);
    });

    it('should go from ASC to DESC', () => {
      const { componentWrapper } = mountMultiEcWithSorting({ sorts: [{ column: 'test-column', direction: SortDirection.ASC }] });
      componentWrapper.vm.$emit('sort', 'test-column');
      expect(componentWrapper.vm.sorts).toEqual([
        { column: 'test-column', direction: SortDirection.DESC },
      ]);
    });

    it('should go from DESC to nothing', () => {
      const { componentWrapper } = mountMultiEcWithSorting({ sorts: [{ column: 'test-column', direction: SortDirection.DESC }] });
      componentWrapper.vm.$emit('sort', 'test-column');
      expect(componentWrapper.vm.sorts).toEqual([]);
    });

    it('should allow sort by multiple columns', () => {
      const { componentWrapper } = mountMultiEcWithSorting();
      componentWrapper.vm.$emit('sort', 'test-column-1');
      componentWrapper.vm.$emit('sort', 'test-column-2');
      expect(componentWrapper.vm.sorts).toEqual([
        { column: 'test-column-1', direction: SortDirection.ASC },
        { column: 'test-column-2', direction: SortDirection.ASC },
      ]);
    });

    it('should update existing sorting when there are multiple column sorted', () => {
      const { componentWrapper } = mountMultiEcWithSorting({
        sorts: [
          { column: 'test-column-1', direction: SortDirection.ASC },
          { column: 'test-column-2', direction: SortDirection.DESC },
        ],
      });

      componentWrapper.vm.$emit('sort', 'test-column-1');
      expect(componentWrapper.vm.sorts).toEqual([
        { column: 'test-column-1', direction: SortDirection.DESC },
        { column: 'test-column-2', direction: SortDirection.DESC },
      ]);

      componentWrapper.vm.$emit('sort', 'test-column-2');
      expect(componentWrapper.vm.sorts).toEqual([
        { column: 'test-column-1', direction: SortDirection.DESC },
      ]);
    });

    it('should preserve sorting order', () => {
      const { componentWrapper } = mountMultiEcWithSorting({
        sorts: [
          { column: 'test-column-1', direction: SortDirection.ASC },
          { column: 'test-column-2', direction: SortDirection.DESC },
          { column: 'test-column-3', direction: SortDirection.ASC },
        ],
      });

      // sorting test-column-1 3 times should move it to bottom of the sorts array
      componentWrapper.vm.$emit('sort', 'test-column-1'); // ASC -> DESC
      componentWrapper.vm.$emit('sort', 'test-column-1'); // DESC -> null
      componentWrapper.vm.$emit('sort', 'test-column-1'); // null -> ASC

      expect(componentWrapper.vm.sorts).toEqual([
        { column: 'test-column-2', direction: SortDirection.DESC },
        { column: 'test-column-3', direction: SortDirection.ASC },
        { column: 'test-column-1', direction: SortDirection.ASC },
      ]);

      // sorting test-column-2 2 times should move it to bottom of the sorts array
      componentWrapper.vm.$emit('sort', 'test-column-2'); // DESC -> null
      componentWrapper.vm.$emit('sort', 'test-column-2'); // null -> ASC

      expect(componentWrapper.vm.sorts).toEqual([
        { column: 'test-column-3', direction: SortDirection.ASC },
        { column: 'test-column-1', direction: SortDirection.ASC },
        { column: 'test-column-2', direction: SortDirection.ASC },
      ]);
    });
  });

  describe('single sort', () => {
    function mountSingleEcWithSorting(props, mountOpts) {
      return mountEcWithSorting({ multiSort: false, ...props }, mountOpts);
    }

    it('should use default sort direction when column is not sorted yet', () => {
      const { componentWrapper } = mountSingleEcWithSorting();
      componentWrapper.vm.$emit('sort', 'test-column');
      expect(componentWrapper.vm.sorts).toEqual([
        { column: 'test-column', direction: SortDirection.ASC },
      ]);
    });

    it('should go from ASC to DESC', () => {
      const { componentWrapper } = mountSingleEcWithSorting({ sorts: [{ column: 'test-column', direction: SortDirection.ASC }] });
      componentWrapper.vm.$emit('sort', 'test-column');
      expect(componentWrapper.vm.sorts).toEqual([
        { column: 'test-column', direction: SortDirection.DESC },
      ]);
    });

    it('should go from DESC to nothing', () => {
      const { componentWrapper } = mountSingleEcWithSorting({ sorts: [{ column: 'test-column', direction: SortDirection.DESC }] });
      componentWrapper.vm.$emit('sort', 'test-column');
      expect(componentWrapper.vm.sorts).toEqual([]);
    });

    it('should not allow sort by multiple columns', () => {
      const { componentWrapper } = mountSingleEcWithSorting();
      componentWrapper.vm.$emit('sort', 'test-column-1');
      componentWrapper.vm.$emit('sort', 'test-column-2');
      expect(componentWrapper.vm.sorts).toEqual([
        { column: 'test-column-2', direction: SortDirection.ASC },
      ]);
    });
  });
});
