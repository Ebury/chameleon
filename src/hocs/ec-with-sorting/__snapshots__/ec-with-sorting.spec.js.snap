// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EcWithSorting should render properly 1`] = `
<div
  defaultsortcycle=",asc,desc"
/>
`;
